feat(ingest): add STT since-based HTTP feeding service using NinJS parser (mainly for Sport results / stats)

0is1 opened the pull request:
* Wire new feeding service app into INSTALLED_APPS
* Implement STTWithSinceHTTPFeedingService (NAME: "stt_since_http") for STT Sport Results API
* Build URL with since=<ISO-8601 Z> from provider.last_updated (fallback to now)
* Auth via Bearer token (auth_token in provider.config), requests.Session with timeout
* Parse response by feeding one NinJS item at a time; supports dict, {items: [...]}, or list payloads
* Normalize timestamps (versioncreated/firstcreated/embargoed) to UTC "Z" to satisfy ninjs datetime parsing
* Adapt categories: map anpa_category → ninjs "service" and normalize service to [{code, name?, scheme?}]

0is1 committed:
feat(ingest): add STT since-based HTTP feeding service using NinJS parser (mainly for Sport results / stats)

* Wire new feeding service app into INSTALLED_APPS
* Implement STTWithSinceHTTPFeedingService (NAME: "stt_since_http") for STT Sport Results API
* Build URL with since=<ISO-8601 Z> from provider.last_updated (fallback to now)
* Auth via Bearer token (auth_token in provider.config), requests.Session with timeout
* Parse response by feeding one NinJS item at a time; supports dict, {items: [...]}, or list payloads
* Normalize timestamps (versioncreated/firstcreated/embargoed) to UTC "Z" to satisfy ninjs datetime parsing
* Adapt categories: map anpa_category → ninjs "service" and normalize service to [{code, name?, scheme?}]

0is1 committed:
feat: always use ninjs parser

petrjasek commented on server/stt/io/feeding_services/stt_http_with_since.py:
```diff
@@ -0,0 +1,358 @@
+                results = self._call_parser(parser, tmp, provider)
+            finally:
+                try:
+                    os.unlink(tmp)
```
imo if you move that call to `_call_parser` inside `with NamedTemporaryFile` block you can keep that default delete behaviour and you don't need to handle the cleanup after

petrjasek commented on server/stt/io/feeding_services/stt_http_with_since.py:
```diff
@@ -0,0 +1,358 @@
+                    normalized.append({"code": el})
+            out["service"] = normalized
+        elif isinstance(out.get("service"), str):
+            out["service"] = [{"code": out.get("service")}]
```
I wonder what's the use for the `service` value? we don't use it in superdesk there we have `anpa_category`

0is1 replied:
Yeah, actually that is completely unnecessary logic for all the "else"-blocks. I removed them in the latest commit. 

But I just noticed that the default ninjs parser expects `service` key that is then mapped to the `anpa_category` (https://github.com/superdesk/superdesk-core/blob/develop/superdesk/io/feed_parsers/ninjs.py#L101-L102) so that's why I added this kind of helper here. Do you have any idea why it's like that?

petrjasek replied:
my bad, didn't realize it will call that ninjs parser on that, makes sense then

0is1 committed:
refactor(stt-http): call parser within NamedTemporary

File context on POSIX (Docker)
- Switch to delete=True and invoke parser inside the with-block
- Remove manual os.unlink cleanup
- add flush+fsync for durability

_adapt_category_for_ninjs
- Remove unnecessary logic

petrjasek commented on server/stt/io/feeding_services/stt_http_with_since.py:
```diff
@@ -0,0 +1,331 @@
+                # - flush(): push Python's buffer to the OS so size/data are visible to a new open()
+                # - fsync(): ask the OS to flush page cache + metadata; helpful on some FS/overlay layers
+                #   (Docker/overlayfs) to avoid stale size/content. Safe but optional for POSIX; drop if perf-critical.
+                f.flush()
```
I think this might require rather `f.seek(0)` after write so that parser car read the file from the beginning
